[css-anchor-position-1] Implement parsing support for position-area

tuankiet65 opened the pull request:
#### 5fd602ab52938ca40524eacce9b980b8adf11b6a
<pre>
[css-anchor-position-1] Implement parsing support for position-area
<a href="https://bugs.webkit.org/show_bug.cgi?id=281289">https://bugs.webkit.org/show_bug.cgi?id=281289</a>
<a href="https://rdar.apple.com/138173544">rdar://138173544</a>

Reviewed by Antti Koivisto.

Implement parsing support for position-area, as specified in the
CSS Anchor Positioning spec:
<a href="https://drafts.csswg.org/css-anchor-position/#position-area">https://drafts.csswg.org/css-anchor-position/#position-area</a>

Note that the spec seems to confuse between the self- keywords
(for example: self-block-end in the syntax, then block-self-end in the
explanation). This patch strictly follows the keywords used in the
syntax definition, which seems to align with Chrome behavior and
WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-allowed-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed-insets-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/property-interpolations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::keywordForPositionAreaSpan):
(WebCore::valueForPositionArea):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::propertyAllowedInPositionTryRule):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp: Added.
(WebCore::CSSPropertyParserHelpers::getKeywordType):
(WebCore::CSSPropertyParserHelpers::typesAreCompatible):
(WebCore::CSSPropertyParserHelpers::typeIsAxisExplicit):
(WebCore::CSSPropertyParserHelpers::typeIsBlockOrXAxis):
(WebCore::CSSPropertyParserHelpers::typeIsInlineOrYAxis):
(WebCore::CSSPropertyParserHelpers::valueForPositionArea):
(WebCore::CSSPropertyParserHelpers::consumePositionArea):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h: Added.
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/rendering/style/PositionArea.cpp: Added.
(WebCore::PositionAreaSpan::PositionAreaSpan):
(WebCore::axisIsBlockOrX):
(WebCore::axisIsInlineOrY):
(WebCore::PositionArea::PositionArea):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/PositionArea.h: Added.
(WebCore::PositionAreaSpan::axis const):
(WebCore::PositionAreaSpan::track const):
(WebCore::PositionAreaSpan::self const):
(WebCore::PositionArea::blockOrXAxis const):
(WebCore::PositionArea::inlineOrYAxis const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareDataChangeRequiresLayout):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialPositionArea):
(WebCore::RenderStyle::positionArea const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setPositionArea):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::positionAreaKeywordToAxis):
(WebCore::Style::positionAreaKeywordToTrack):
(WebCore::Style::positionAreaKeywordToSelf):
(WebCore::Style::positionAreaExpandKeyword):
(WebCore::Style::positionAreaOppositeAxis):
(WebCore::Style::BuilderConverter::convertPositionArea):

Canonical link: <a href="https://commits.webkit.org/290113@main">https://commits.webkit.org/290113@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2015de8ade12008835c3bf6db30393e751a491e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94003 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39789 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91080 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16737 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26269 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6831 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/48959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38897 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76942 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95840 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16209 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11850 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76458 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/76764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18914 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21155 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9292 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16223 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21534 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15964 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17745 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
```diff
@@ -0,0 +1,216 @@
+namespace WebCore {
+namespace CSSPropertyParserHelpers {
+
+enum class Type {
```
```suggestion
enum class Type : uint8_t {
```

smfr commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+*/
+
+// 6 values, 3 bits
+enum class PositionAreaAxis {
```
`: uint8_t`

You're not putting this in a bitfield anywhere, so the comment about bits is not applicable.

tuankiet65 replied:
Yeah this is when I though I'd be able to compress everything down to one byte; turns out the minimum size is 4 bytes anyway. Will remove.

smfr commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+};
+
+// 6 values, 3 bits
+enum class PositionAreaTrack {
```
Ditto

smfr commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+    SpanAll,
+};
+
+enum class PositionAreaSelf { No, Yes };
```
`: bool`

fantasai commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3163,6 +3163,40 @@ static Ref<CSSValue> valueForAnchorName(const Vector<Style::ScopedName>& scopedN
+        );
+    };
+
+    if (!positionArea)
+        return CSSPrimitiveValue::create(CSSValueNone);
```
Move this short-circuit first in the function.

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1937,6 +1937,8 @@ void RenderStyle::conservativelyCollectChangedAnimatableProperties(const RenderS
             changingProperties.m_properties.set(CSSPropertyAnchorName);
         if (first.positionAnchor != second.positionAnchor)
             changingProperties.m_properties.set(CSSPropertyPositionAnchor);
+        if (first.positionArea != second.positionArea)
```
You need to do something for RenderStyle::diff() too (presumably you'd trigger layout).

fantasai commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
```diff
@@ -0,0 +1,216 @@
+    SelfLogicalInline,
+    Ambiguous,
+    SelfAmbiguous,
+    Axesless
```
Maybe `Axisless` is better? :)

fantasai commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
```diff
@@ -0,0 +1,216 @@
+    if (dim1Type == Type::PhysicalY || dim1Type == Type::LogicalInline || dim1Type == Type::SelfLogicalInline || dim2Type == Type::PhysicalX || dim2Type == Type::LogicalBlock || dim2Type == Type::SelfLogicalBlock)
+        return CSSValuePair::create(dim2.releaseNonNull(), dim1.releaseNonNull());
+
+    return CSSValuePair::create(dim1.releaseNonNull(), dim2.releaseNonNull());
```
The job of the parser is to do the parsing, so how about we make this the _only_ place that we handle this ordering and optionality logic?

Make this function _always_ return two values (unless it's `none`), with a guaranteed axis ordering for a given type of axes.

tuankiet65 replied:
We discussed this elsewhere, but the `CSSValue` returned by this function is used in specified value serialization; the test expects that this value is simplified, so we can't make it always return a pair.

fantasai commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+
+class PositionArea {
+public:
+    PositionArea(PositionAreaSpan span1, PositionAreaSpan span2);
```
Require the user to pass in the values in the correct axis ordering, and name the variables accordingly.

fantasai commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+    PositionArea(PositionAreaSpan span1, PositionAreaSpan span2);
+
+    static PositionArea fromKeyword(CSSValueID);
+    static PositionArea fromKeywordPair(CSSValueID kw1, CSSValueID kw2);
```
Same here.

fantasai commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+    static PositionArea fromKeywordPair(CSSValueID kw1, CSSValueID kw2);
+
+    PositionAreaSpan span1() const { return m_span1; }
+    PositionAreaSpan span2() const { return m_span2; }
```
This can also be named accordingly, e.g. `blockOrXAxis()` and `inlineOrYAxis()`.

tuankiet65 replied:
Fixed.

fantasai commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -0,0 +1,113 @@
+
+private:
+    PositionAreaSpan m_span1;
+    PositionAreaSpan m_span2;
```
Same.

fantasai commented on Source/WebCore/rendering/style/PositionArea.cpp:
```diff
@@ -0,0 +1,503 @@
+
+    // Keyword is axis ambiguous, it's repeated
+    return PositionArea::fromKeywordPair(kw, kw);
+}
```
If the parser always returns two values, then we don't need this function.

tuankiet65 replied:
I said before that the parser can't always return two values because of how serialization of `position-area` specified value is, so unfortunately we have to keep this.

fantasai commented on Source/WebCore/rendering/style/PositionArea.cpp:
```diff
@@ -0,0 +1,503 @@
+        { *kw1Axis, keywordToTrack(kw1), keywordToSelf(kw1) },
+        { *kw2Axis, keywordToTrack(kw2), keywordToSelf(kw2) }
+    };
+}
```
If the parser is guaranteeing two values in the correct order, you can simplify your logic and all of your helper functions down to a single switch/case that constructs the PositionSpan object for each axis span from a keyword. Something like:

```
PositionSpan PositionSpan::fromKeyword(CSSValueID keyword)
{
switch (keyword) {
    case CSSValueLeft:
        return { Horizontal, Start, No };
    case CSSValueSpanLeft:
        return { Horizontal, SpanStart, No };
    case CSSValueRight:
        return { Horizontal, SpanEnd, No };
 etc.
}

PositionArea(CSSValueID blockOrXKeyword, CSSValueID inlineOrYKeyword)
: m_blockOrXSpan(PositionSpan::fromKeyword(blockOrXKeyword))
, m_inlineOrYSpan(PositionSpan::fromKeyword(inlineOrYKeyword))
{ }

tuankiet65 replied:
Some keywords like `start`, `end`, `center` doesn't have an axis; its axis is the opposite of the axis of the other keyword. If we were to write a function that maps keyword => PositionAreaSpan, then it must also take the other keyword, in case it needs to do axis resolution. I think the current solution is better: we have 3 separate functions to determine axis, track and self. Then we could handle axis resolution elsewhere.

tuankiet65 replied:
Also the parser now guarantees if the parser returns a pair, then it's in the correct order (block/X axis first, then inline/Y)

mdubet commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -470,6 +471,7 @@ inline Length RenderStyle::initialPerspectiveOriginY() { return { 50.0f, LengthT
 constexpr PointerEvents RenderStyle::initialPointerEvents() { return PointerEvents::Auto; }
 constexpr PositionType RenderStyle::initialPosition() { return PositionType::Static; }
 inline std::optional<Style::ScopedName> RenderStyle::initialPositionAnchor() { return { }; }
+inline std::optional<PositionArea> RenderStyle::initialPositionArea() { return std::nullopt; }
```
`return { };`

mdubet commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2356,6 +2357,294 @@ inline std::optional<ScopedName> BuilderConverter::convertPositionAnchor(Builder
+}
+
+// Expand a one keyword position-area to the equivalent keyword pair value.
+static std::pair<CSSValueID, CSSValueID> positionAreaExpandKeyword(CSSValueID kw)
```
We generally use full word (kw -> keyword)

tuankiet65 committed:
[css-anchor-position-1] Implement parsing support for position-area
https://bugs.webkit.org/show_bug.cgi?id=281289
rdar://138173544

Reviewed by Antti Koivisto.

Implement parsing support for position-area, as specified in the
CSS Anchor Positioning spec:
https://drafts.csswg.org/css-anchor-position/#position-area

Note that the spec seems to confuse between the self- keywords
(for example: self-block-end in the syntax, then block-self-end in the
explanation). This patch strictly follows the keywords used in the
syntax definition, which seems to align with Chrome behavior and
WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-allowed-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed-insets-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/property-interpolations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::keywordForPositionAreaSpan):
(WebCore::valueForPositionArea):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::propertyAllowedInPositionTryRule):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp: Added.
(WebCore::CSSPropertyParserHelpers::getKeywordType):
(WebCore::CSSPropertyParserHelpers::typesAreCompatible):
(WebCore::CSSPropertyParserHelpers::typeIsAxisExplicit):
(WebCore::CSSPropertyParserHelpers::typeIsBlockOrXAxis):
(WebCore::CSSPropertyParserHelpers::typeIsInlineOrYAxis):
(WebCore::CSSPropertyParserHelpers::valueForPositionArea):
(WebCore::CSSPropertyParserHelpers::consumePositionArea):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h: Added.
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/rendering/style/PositionArea.cpp: Added.
(WebCore::PositionAreaSpan::PositionAreaSpan):
(WebCore::axisIsBlockOrX):
(WebCore::axisIsInlineOrY):
(WebCore::PositionArea::PositionArea):
(WebCore::operator<<):
* Source/WebCore/rendering/style/PositionArea.h: Added.
(WebCore::PositionAreaSpan::axis const):
(WebCore::PositionAreaSpan::track const):
(WebCore::PositionAreaSpan::self const):
(WebCore::PositionArea::blockOrXAxis const):
(WebCore::PositionArea::inlineOrYAxis const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareDataChangeRequiresLayout):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialPositionArea):
(WebCore::RenderStyle::positionArea const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setPositionArea):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::positionAreaKeywordToAxis):
(WebCore::Style::positionAreaKeywordToTrack):
(WebCore::Style::positionAreaKeywordToSelf):
(WebCore::Style::positionAreaExpandKeyword):
(WebCore::Style::positionAreaOppositeAxis):
(WebCore::Style::BuilderConverter::convertPositionArea):

Canonical link: https://commits.webkit.org/290113@main
